engine/src/main/java/org/hibernate/validator/internal/engine/valueextraction/ValueExtractorManager.java

```diff
@@ -347,6 +351,124 @@ private static boolean isClassPresent(String className, boolean fallbackOnTCCL)
 		}
 	}
 
+	private final class ValueExtractorCache {
+
+		private final ConcurrentHashMap<ValueExtractorCacheKey, Set<ValueExtractorDescriptor>> valueExtractorCache;
+		private final Set<ValueExtractorCacheKey> nonContainerTypes;
+
+		public ValueExtractorCache() {
+			this.valueExtractorCache = new ConcurrentHashMap<>();
+			this.nonContainerTypes = new HashSet<>();
+		}
+
+		public Set<ValueExtractorDescriptor> getValueExtractors( Set<ValueExtractorDescriptor> valueExtractorCandidates,
+							Class<?> valueType ) {
+			ValueExtractorCacheKey cacheKey = new ValueExtractorCacheKey( valueType, /*something here*/ null );
+
+			if ( nonContainerTypes.contains( cacheKey ) ) {
```

Comment:
You can avoid using the cache key here. If it's a non container type, the whole type is not a container. This means that in the simple case where it's not a container, we avoid instantiating the cache key.